crates/engine/tree/src/tree/precompile_cache.rs

```diff
@@ -0,0 +1,221 @@
+//! Contains a precompile cache that is backed by a moka cache.
+
+use alloy_primitives::Bytes;
+use reth_evm::precompiles::{DynPrecompile, Precompile};
+use revm::precompile::{PrecompileError, PrecompileResult};
+use std::sync::Arc;
+
+/// Cache for precompiles, for each input stores the result.
+pub type PrecompileCache =
+    mini_moka::sync::Cache<CacheKey, CacheEntry, alloy_primitives::map::DefaultHashBuilder>;
+
+/// Create a new [`PrecompileCache`].
+pub fn create_precompile_cache() -> PrecompileCache {
+    mini_moka::sync::CacheBuilder::new(100_000)
+        .build_with_hasher(alloy_primitives::map::DefaultHashBuilder::default())
+}
+
+/// Cache key, just input for each precompile.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct CacheKey(alloy_primitives::Bytes);
+
+/// Combined entry containing both the result and gas bounds.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct CacheEntry {
+    /// The actual result of executing the precompile.
+    result: PrecompileResult,
+    /// Observed gas limit above which the precompile does not fail with out of gas.
+    upper_gas_limit: u64,
+    /// Observed gas limit below which the precompile fails with out of gas.
+    lower_gas_limit: u64,
+}
+
+/// A cache for precompile inputs / outputs.
+///
+/// This assumes that the precompile is a standard precompile, as in `StandardPrecompileFn`, meaning
+/// its inputs are only `(Bytes, u64)`.
+///
+/// NOTE: This does not work with "context stateful precompiles", ie `ContextStatefulPrecompile` or
+/// `ContextStatefulPrecompileMut`. They are explicitly banned.
+#[derive(Debug)]
+pub(crate) struct CachedPrecompile {
+    /// Cache for precompile results and gas bounds.
+    cache: Arc<PrecompileCache>,
+    /// The precompile.
+    precompile: DynPrecompile,
+    /// Cache metrics.
+    metrics: CachedPrecompileMetrics,
+}
+
+impl CachedPrecompile {
+    /// `CachedPrecompile` constructor.
+    pub(crate) fn new(precompile: DynPrecompile, cache: Arc<PrecompileCache>) -> Self {
+        Self { precompile, cache, metrics: Default::default() }
+    }
+
+    pub(crate) fn wrap(precompile: DynPrecompile, cache: Arc<PrecompileCache>) -> DynPrecompile {
+        |data: &Bytes, gas_limit: u64| -> PrecompileResult {
+            Self::new(precompile, cache).call(data, gas_limit)
+        }
```

Comment:
this isn't quite right, because this is the precompile fn, which internally now creates a new `CachedPrecompile` and invokes it

the conversion we need here `Self::new(precompile, cache).into()`